### PR TITLE
Make datanode load statslog lazy if SkipBFStatsLog is true

### PR DESCRIPTION
**Pull request by congqixia**
See also #22994 
Related to #23477
/kind improvement

### CODE REVIEW

File: internal/datanode/channel_meta.go [hunk 1]
```diff
@@ -294,13 +311,39 @@ func (c *ChannelMeta) filterSegments(partitionID UniqueID) []*Segment {
 }
 
 func (c *ChannelMeta) InitPKstats(ctx context.Context, s *Segment, statsBinlogs []*datapb.FieldBinlog, ts Timestamp) error {
+	if Params.DataNodeCfg.SkipBFStatsLoad {
+		// mark segment lazy loading
+		s.setLoadingLazy(true)
+		go func() {
```
Comment (xiaofan-luan) [inline internal/datanode/channel_meta.go:317]: didn't really make any sense to me
Why we go the function out but mean time we use a pool?

File: internal/datanode/channel_meta.go [hunk 2]
```diff
@@ -294,13 +311,39 @@ func (c *ChannelMeta) filterSegments(partitionID UniqueID) []*Segment {
...
+			c.workerPool.Submit(func() (any, error) {
+				stats, err := c.loadStats(context.Background(), s.segmentID, s.collectionID, statsBinlogs, ts)
+				if err != nil {
+					log.Warn("failed to lazy load statslog for segment", zap.Int64("segmentID", s.segmentID), zap.Int64("collectionID", s.collectionID))
+					return nil, err
```
Comment (xiaofan-luan) [inline internal/datanode/channel_meta.go:322]: and also what happens if lazy load failed?
I think we should retry for a very long time. until some time we will just abort the service or release the channle

File: internal/datanode/channel_meta.go [hunk 3]
```diff
@@ -111,13 +114,26 @@ type ChannelMeta struct {
 
 	metaService  *metaService
 	chunkManager storage.ChunkManager
+	workerPool   *concurrency.Pool
+}
+
+type lazyLoadTask struct {
```
Comment (xiaofan-luan) [inline internal/datanode/channel_meta.go:120]: lazyLoadTask seems not to be used at all

Comment (congqixia) [reply internal/datanode/channel_meta.go]: removed

File: internal/datanode/channel_meta.go (hunk 1 above)
Comment (congqixia) [reply internal/datanode/channel_meta.go]: removed

File: internal/datanode/channel_meta.go (hunk 2 above)
Comment (xiaofan-luan) [reply internal/datanode/channel_meta.go]: But what it an io error or a S3 rate limit?

File: internal/datanode/channel_meta.go (hunk 1 above)
Comment (congqixia) [reply internal/datanode/channel_meta.go]: Recalled why add a goroutine to submit task into the work pool: if the worker pool is full, the Submit call will block here until there are some available worker.

File: internal/datanode/channel_meta.go (hunk 2 above)
Comment (congqixia) [reply internal/datanode/channel_meta.go]: Added re-submit logic and error type check logic here

File: internal/datanode/channel_meta.go [hunk 4]
```diff
@@ -294,13 +311,68 @@ func (c *ChannelMeta) filterSegments(partitionID UniqueID) []*Segment {
...
+					c.submitLoadStatsTask(s, statsBinlogs, ts)
+				}
+				return nil, err
+			}
+			// get segment lock here
+			// it's ok that segment is dropped here
+			c.segMu.Lock()
+			defer c.segMu.Unlock()
+			s.historyStats = append(s.historyStats, stats...)
+			s.setLoadingLazy(false)
```
Comment (xiaofan-luan) [inline internal/datanode/channel_meta.go:350]: need a log when successfully load the segment

File: internal/datanode/channel_meta.go [hunk 5]
```diff
@@ -294,13 +311,68 @@ func (c *ChannelMeta) filterSegments(partitionID UniqueID) []*Segment {
...
+		return
+	}
+	// do submitting in a goroutine in case of task pool is full
+	go func() {
+		c.workerPool.Submit(func() (any, error) {
+			stats, err := c.loadStats(context.Background(), s.segmentID, s.collectionID, statsBinlogs, ts)
+			if err != nil {
+				// TODO if not retryable, add rebuild statslog logic
+				log.Warn("failed to lazy load statslog for segment", zap.Error(err))
+				if c.retryableLoadError(err) {
```
Comment (xiaofan-luan) [inline internal/datanode/channel_meta.go:339]: should we take of close here?

File: internal/datanode/channel_meta.go (hunk 4 above)
Comment (congqixia) [reply internal/datanode/channel_meta.go:350]: done

File: internal/datanode/channel_meta.go (hunk 5 above)
Comment (congqixia) [reply internal/datanode/channel_meta.go:339]: The `closed` check  is at the beginning of this resubmit function

**Pull request merged**